examples/gno.land/p/demo/grc/grc721/grc721.gno

```diff
@@ -0,0 +1,261 @@
+	_ = nft.mustBeOwned(tokenId)
+	return nft.getApproved(tokenId)
+}
+
+func (nft Token) IsApprovedForAll(owner, operator std.Address) bool {
+	approved, exists := nft.operatorApprovals.Get(operatorKey(owner, operator))
+	if !exists || approved == false {
+		return false
+	}
+
+	return true
+}
+
+func (nft Token) TokenURI(tokenId string) string {
+	nft.mustBeOwned(tokenId)
+	uri, exists := nft.tokenURIs.Get(tokenId)
+	if !exists {
+		return ""
+	}
+
+	return uri.(string)
+}
+
+func (nft Token) SetTokenURI(tokenId string, tokenURI string) string {
+	nft.tokenURIs.Set(tokenId, tokenURI)
+	return tokenURI
+}
+
+func (nft Token) Mint(to std.Address, tokenId string) {
+	mustBeValid(to)
+	prevOwner := nft.update(to, tokenId, emptyAddress)
+	if prevOwner != emptyAddress {
+		str := ufmt.Sprintf("GRC721: token with id %s has already been minted", tokenId)
+		panic(str)
+	}
+}
+
+func (nft Token) Burn(tokenId string) {
+	prevOwner := nft.update(emptyAddress, tokenId, emptyAddress)
+
+	if prevOwner == emptyAddress {
+		str := ufmt.Sprintf("GRC721: Token with ID %s does not exist", tokenId)
+		panic(str)
+	}
+}
+
+// Helpers
+func (nft Token) requireOwner(caller std.Address, tokenId string) {
+	if caller != nft.mustBeOwned(tokenId) {
+		panic("GRC721: not owner")
+	}
+}
+
+func (nft Token) getApproved(tokenId string) std.Address {
+	approved, exists := nft.tokenApprovals.Get(tokenId)
+	if !exists {
+		return "" // panic instead?
+	}
+
+	return approved.(std.Address)
+}
+
+// mustBeValid panics if the given address is not valid
+func mustBeValid(address std.Address) {
+	if !address.IsValid() {
+		err := ufmt.Sprintf("GRC721: invalid address %s", address)
+		panic(err)
+	}
+}
+
+// mustBeOwned panics if token is not owned by an address (does not exist)
+// If the token is owned, mustBeOwned returns the owner of the token
+func (nft Token) mustBeOwned(tokenId string) std.Address {
+	owner, exists := nft.owners.Get(tokenId)
+	if !exists {
+		err := ufmt.Sprintf("GRC721: token with ID %s does not exist", tokenId)
+		panic(err)
+	}
+
+	return owner.(std.Address)
+}
+
+// checkAuthorized checks if spender is authorized to spend specified token on behalf of owner
+// Panics if token doesn't exist, or if spender is not authorized in any way
+func (nft Token) checkAuthorized(owner, spender std.Address, tokenId string) {
+	_ = nft.mustBeOwned(tokenId)
+
+	if !nft.isAuthorized(owner, spender, tokenId) {
+		str := ufmt.Sprintf("GRC721: %s is not authorized for %s", spender, tokenId)
+		panic(str)
+	}
+}
+
+// isAuthorized returns if the spender is authorized to transfer the specified token
+// Assumes addresses are valid and the token exists
+func (nft Token) isAuthorized(owner, spender std.Address, tokenId string) bool {
+	return owner == spender ||
+		nft.IsApprovedForAll(owner, spender) ||
+		nft.getApproved(tokenId) == owner
+}
+
+func (nft Token) update(to std.Address, tokenId string, auth std.Address) std.Address {
```

Comment:
```suggestion
func (nft *Token) update(to std.Address, tokenId string, auth std.Address) std.Address {
```

was the code currently working?
because it shouldn't.

Reply:
It seems that this was working mainly because I use pointers to the AVL storage. Tests were also passing. I'll see to either use non-pointers for AVL trees or just add pointer receivers.

Reply:
Ended up just making the function take receiver pointers.